server/controllers/files.py

```diff
@@ -0,0 +1,45 @@
+import logging
+
+import libcloud
+from flask import Blueprint, redirect, Response, abort
+from flask_login import login_required, current_user
+
+from server.extensions import storage
+from server.models import ExternalFile
+
+files = Blueprint('files', __name__)
+
+logger = logging.getLogger(__name__)
+
+@files.route("/files/<hashid:file_id>")
+@login_required
+def file_url(file_id):
+    ext_file = ExternalFile.query.filter_by(id=file_id, deleted=False).first()
+    if not ext_file or not ExternalFile.can(ext_file, current_user, 'download'):
+        abort(404)
+
+    try:
+        storage_obj = ext_file.object
+    except libcloud.common.types.InvalidCredsError as e:
+        logger.warning("Could not get file {0} - {1}".format(file_id, ext_file.filename),
+                       exc_info=True)
+        storage_obj = None
+
+    if storage_obj is None:
+        abort(404, "File does not exist")
+
+    # Do not use .download_url for local storage.
+    if "local" in storage.driver.name.lower():
+        def streamer(stream):
+            for data in stream:
+                yield data
```

Comment:
In the sense that it'll never be executed in production? I could see some one running their own deployment on heroku or their own server and then this would be useful.